### PR TITLE
feat(tracing): add OpenTelemetry traces with gen_ai.* and tool spans

### LINKED ISSUE #19: [Feature]: otel-llm / openllmetry compatible spans
### Is your feature request related to a problem?

Love that you've created this plugin to actually allow exporting OTEL signals.

Are there any plans to emit traces and spans? Having dug in to the opencode hooks available, I'm not sure how feasible it would be to actually report fully valid spans using a plugin (the native integration makes pushing session and prompt span context a bit easier e.g. https://github.com/anomalyco/opencode/pull/6629)  but from what I can seem to infer, OTEL support isn't a top priority for the core team.

### Describe the solution you'd like

I want to be able to collect full traces so I can build evals, scorers, and test if our internal coding harness improves things for the team.

### Describe alternatives you've considered

_No response_

### Telemetry type

Other

### Additional context

_No response_

### Checklist

- [x] I have searched existing issues and discussions for similar feature requests
- [x] I have checked the changelog to see if this already exists

### PR DESCRIPTION
**Pull request by dialupdisaster**
Closes #19

## What this adds

Full OTel trace instrumentation across session, LLM message, and tool lifecycles, exported via OTLP/gRPC alongside the existing metrics and logs.

### Trace structure

```
opencode.session (root span)
  └─ gen_ai.chat (SpanKind.CLIENT — per assistant message)
  └─ opencode.tool <name> (SpanKind.INTERNAL — per tool call)
  └─ opencode.tool <name>
  ...
```

Subagent sessions (`parentID` present) are nested under their parent session span.

### Spans and attributes

**`opencode.session`**
- `session.id`, `session.is_subagent`, `project.id`
- On idle: `session.total_tokens`, `session.total_cost_usd`, `session.total_messages`
- Status: `OK` on idle, `ERROR` on session.error

**`gen_ai.chat`** (OpenLLMetry compatible)
- `gen_ai.system` (provider), `gen_ai.request.model`, `session.id`
- On completion: `gen_ai.usage.input_tokens`, `gen_ai.usage.output_tokens`, `gen_ai.usage.reasoning_tokens`, `gen_ai.usage.cache_read_tokens`, `gen_ai.usage.cache_creation_tokens`, `gen_ai.response.finish_reason`, `cost_usd`, `duration_ms`
- Status: `OK` or `ERROR`

**`opencode.tool <name>`**
- `tool.name`, `session.id`
- On success: `tool.result_size_bytes`
- On error: `tool.error`
- Status: `OK` or `ERROR`

### Edge cases handled

| Edge case | Handling |
|---|---|
| Orphaned tool spans (session ends before tool completes) | Swept with `ERROR` status in `sweepSession` on `session.idle` and `session.error` |
| Orphaned message spans (session ends before message completes) | Same sweep, `ERROR` status |
| Out-of-order tool events (`completed` arrives without prior `running`) | Best-effort span started and immediately ended |
| Subagent with missing parent span | Falls back to a root span, no error |
| `session.error` with undefined `sessionID` | Session span end skipped safely |

### Known limitations (documented in #19)

- Prompt/completion content is not available in spans — the plugin API does not expose message bodies
- Span context cannot be injected into the outbound LLM HTTP call — requires native core instrumentation

## Changes

| File | Change |
|---|---|
| `package.json` | Add `@opentelemetry/sdk-trace-base`, `@opentelemetry/exporter-trace-otlp-grpc` |
| `src/otel.ts` | `BasicTracerProvider` + `BatchSpanProcessor` + `OTLPTraceExporter`; `setupOtel` returns `tracerProvider` |
| `src/types.ts` | `PendingToolSpan` gains `span: Span`; `HandlerContext` gains `tracer`, `sessionSpans`, `messageSpans` |
| `src/index.ts` | Wires tracer and span maps into context; calls `startMessageSpan` on incomplete `message.updated`; includes `tracerProvider.shutdown()` |
| `src/handlers/session.ts` | Session span lifecycle; subagent parent linking; orphan sweep |
| `src/handlers/message.ts` | `startMessageSpan()` export; LLM span lifecycle with `gen_ai.*` attrs; tool span lifecycle; out-of-order handling |
| `tests/helpers.ts` | `SpySpan`, `SpyTracer`, `makeTracer()` spy infrastructure |
| `tests/handlers/spans.test.ts` | 34 new tests covering all span behaviours and edge cases |

**185 tests pass, 0 type errors.**

## Summary by CodeRabbit

* **New Features**
  * Added distributed tracing for sessions, assistant messages, and tool calls with lifecycle tracking and cleanup.
  * New config option to disable specific trace types at runtime.

* **Chores**
  * Added OpenTelemetry tracing packages to dependencies.

* **Tests**
  * Added comprehensive tests and test helpers to validate tracing behavior, lifecycle, out-of-order handling, and disablement modes.

### CODE REVIEW

File: src/handlers/session.ts [hunk 1]
```diff
@@ -28,11 +55,23 @@ function sweepSession(sessionID: string, ctx: HandlerContext) {
...
+      ctx.pendingToolSpans.delete(key)
+    }
+  }
+  for (const [msgID, span] of ctx.messageSpans) {
+    const attrs = (span as unknown as { attributes?: Record<string, unknown> }).attributes
+    if (attrs?.["session.id"] === sessionID) {
+      span.setStatus({ code: SpanStatusCode.ERROR, message: "session ended before message completed" })
+      span.end()
+      ctx.messageSpans.delete(msgID)
+    }
```
[the remark this answers is not in the document]
Comment (dialupdisaster) [reply src/handlers/session.ts]: This was actually addressed in the previous commit (`29bd7e6`). The `messageSpans` map is now keyed by `${sessionID}:${messageID}`, and the sweep logic matches on the key prefix (`key.startsWith(\`${sessionID}:\`)`) rather than casting into span internals. Ownership is tracked directly via the composite key.

File: src/util.ts [hunk 2]
```diff
@@ -30,6 +30,14 @@ export function isMetricEnabled(name: string, ctx: { disabledMetrics: Set<string
   return !ctx.disabledMetrics.has(name)
 }
 
+/**
+ * Returns `true` if the trace type is not in the disabled set.
+ * Valid names are `"session"`, `"llm"`, and `"tool"`.
+ */
+export function isTraceEnabled(name: string, ctx: { disabledTraces: Set<string> }): boolean {
+  return !ctx.disabledTraces.has(name)
+}
```
[the remark this answers is not in the document]
Comment (dialupdisaster) [reply src/util.ts:39]: Good catch on the surface, but in practice this invariant is already guaranteed. If `OPENCODE_ENABLE_TELEMETRY` is unset, `config.enabled` is false, and the plugin returns an empty `{}` object immediately at `src/index.ts:44`. The event handlers are never registered with opencode, so `isTraceEnabled` can never be called when the global enabled flag is false. Threading `enabled` through the `HandlerContext` to double-check it in the utility function would add complexity for an unreachable state.
